gh-99069: Consolidate checks for static_assert

jmroot opened the pull request:
Several platforms don't define the static_assert macro despite having
compiler support for the _Static_assert keyword. The macro needs to be
defined since it is used unconditionally in the Python code. So it
should always be safe to define it if undefined and not in C++11 (or
later) mode.

Hence, remove the checks for particular platforms or libc versions,
and just define static_assert anytime it needs to be defined but isn't.
That way, all platforms that need the fix will get it, regardless of
whether someone specifically thought of them.

Also document that certain macOS versions are among the platforms that
need this.


<!-- gh-issue-number: gh-91731 -->
* Issue: gh-91731
<!-- /gh-issue-number -->


<!-- gh-issue-number: gh-99069 -->
* Issue: gh-99069
<!-- /gh-issue-number -->

jmroot committed:
gh-99069: Consolidate checks for static_assert

Several platforms don't define the static_assert macro despite having
compiler support for the _Static_assert keyword. The macro needs to be
defined since it is used unconditionally in the Python code. So it
should always be safe to define it if undefined and not in C++11 (or
later) mode.

Hence, remove the checks for particular platforms or libc versions,
and just define static_assert anytime it needs to be defined but isn't.
That way, all platforms that need the fix will get it, regardless of
whether someone specifically thought of them.

Also document that certain macOS versions are among the platforms that
need this.

jmroot committed:
Workaround for MSVC

vstinner commented on Include/pymacro.h:
```diff
@@ -3,20 +3,20 @@
-// compiler support (gcc >= 4.6) and is called _Static_assert.
+// macOS <= 10.10 doesn't define static_assert in assert.h at all despite
+// having C11 compiler support.
+
```
Can you please use ``//`` for empty lines as well, so it's easy to see the scope of the whole comment at once?

jmroot committed:
Indicate blank lines are part of the comment

pablogsal commented on Include/pymacro.h:
```diff
@@ -3,20 +3,20 @@
-     && !defined(static_assert))
+#if !defined(static_assert) && !defined(_MSC_VER) \
+     && !(defined(__cplusplus) && __cplusplus >= 201103L)
 #  define static_assert _Static_assert
```
Clang will need also a check for `__has_feature(c_static_assert)` here because it may be that `_Static_assert` is not defined

jmroot replied:
Is that really necessary given that C11 is required and  `_Static_assert` is part of that standard? Or is the concern that C99 projects might try to use `static_assert` if it's defined?

pablogsal replied:
Some compiler versions don't implement the full standard (Microsoft compiler is the obvious one but there are more) and some old versions of clang / libc may not implement it either. I also don't know what clang does on AIX for instance, where half of these things are just missing

vstinner replied:
AIX is not longer offiically supported by Python: https://peps.python.org/pep-0011/ If someone gets an issue on AIX and proposes a PR, I'm open to review it. But I prefer to not make assumptions on how AIX should or should not behave.

vstinner replied:
Does it work to express the test as ``!(defined(__has_feature) && __has_feature(c_static_assert))``? If no, maybe we should consider adding _Py__has_feature(), based on this existing macro:

```c
// Preprocessor check for a builtin preprocessor function. Always return 0
// if __has_builtin() macro is not defined.
//
// __has_builtin() is available on clang and GCC 10.
#ifdef __has_builtin
#  define _Py__has_builtin(x) __has_builtin(x)
#else
#  define _Py__has_builtin(x) 0
#endif
```

If I recall correctly, it doesn't work to write ``#if defined(__has_builtin) && __has_builtin(xxx)``.

jmroot replied:
`_Static_assert` is a language keyword, so libc shouldn't be a factor, just the compiler. Would replacing the `__cplusplus` checks with `defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L` be sufficient? Seems like the worst that could happen in that case is you go from an error saying "static_assert undefined" to one saying "_Static_assert undefined".

jmroot replied:
> Does it work to express the test as `!(defined(__has_feature) && __has_feature(c_static_assert))`?

It wouldn't, because the preprocessor evaluates all macros on a line before doing anything else, so short-circuit evaluation doesn't prevent it from trying to evaluate the possibly-undefined thing on the right.

pablogsal replied:
> AIX is not longer offiically supported by Python: https://peps.python.org/pep-0011/ If someone gets an issue on AIX and proposes a PR, I'm open to review it. But I prefer to not make assumptions on how AIX should or should not behave.

I know, that was just an example

jmroot committed:
Restrict static_assert define to gcc/clang C11+

vstinner commented on Include/pymacro.h:
```diff
@@ -3,20 +3,20 @@
+//
+// MSVC makes static_assert a keyword, contrary to the C standard.
+//
 // In C++ 11 static_assert is a keyword, redefining is undefined behaviour.
```
I'm now worried by this comment. You removed ``!(defined(__cplusplus) && __cplusplus >= 201103L)``: did you do it on purpose? If yes, why keeping this scary comment?

jmroot replied:
Checking `__STDC_VERSION__` ensures that we are building as C, not C++.

vstinner replied:
Can please explain it in the comment?

jmroot committed:
Add more explanation of __STDC_VERSION__ check

vstinner commented on Include/pymacro.h:
```diff
@@ -3,20 +3,21 @@
-     && (__GLIBC__ < 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ <= 16)) \
-     && !(defined(__cplusplus) && __cplusplus >= 201103L) \
-     && !defined(static_assert))
+// So only define if building as C, not C++.
```
```suggestion
// So only define if building as C (if __STDC_VERSION__ is defined), not C++.
```

vstinner committed:
Update Include/pymacro.h

vstinner commented on Include/pymacro.h:
```diff
@@ -3,20 +3,21 @@
-     && !(defined(__cplusplus) && __cplusplus >= 201103L) \
-     && !defined(static_assert))
+// So only define if building as C (if __STDC_VERSION__ is defined), not C++.
+#if !defined(static_assert) && defined(__GNUC__) \
```
```suggestion
#if !defined(static_assert) && (defined(__GNUC__) || defined(__clang__)) \
```

vstinner reviewed (commented):
For me, it's strange to check for __GNUC__ with clang. The first code was added for FreeBSD which uses clang.

vstinner committed:
Update Include/pymacro.h

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

jmroot committed:
C2x compatibility for static_assert

The C2x draft (currently expected to become C23) makes static_assert
a keyword to match C++. So only define the macro for up to C17.
